### PR TITLE
Create sieve_of_eratosthenes.py

### PR DESCRIPTION
**Pull request by vedangmehta**
I want to contribute to create a new section of Maths algorithms.

### CODE REVIEW

**Comment by prakhar1989** [comment]
@vedangmehta: Thanks for the PR. The idea to create a new section for Mathematics based algorithms sounds great but before we do that can you suggest at least 3-4 more such algorithms that would fit? For one odd cases, we can add this to `misc` section.

Also, please add a few [unittests](https://docs.python.org/2/library/unittest.html) for your code. Refer to any of the files in the `tests/` folder to see how that is implemented. Lastly, I would suggest adding a [docstring](https://en.wikipedia.org/wiki/Docstring#Python) on the top of your file with a link to the algorithm so that others can make sense of how the code works. 

I'm in the process of cleaning this repository up and I'd prefer that all new PRs conform to this standard. Please feel free to raise any queries that you have.

File: Maths/sieve_of_eratosthenes.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from math import sqrt,ceil
+
+def sieve_of_eratosthenes(n):
+    bool_array = [True] * (n+1)
+    bool_array[0] = False
+    bool_array[1] = False
+    prime_array = []
+    upper_bound = ceil(sqrt(n))
+    for i in range(2,upper_bound):
+        if bool_array[i] == True:
```
Comment (prakhar1989) [inline Maths/sieve_of_eratosthenes.py]: This can just be
```python
if bool_array[i]
```

File: Maths/sieve_of_eratosthenes.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+            for j in range(i*i,n+1,i):
+                bool_array[j] = False
+    for i in range(n+1):
+        if bool_array[i] == True:
```
Comment (prakhar1989) [inline Maths/sieve_of_eratosthenes.py]: Same here

**Comment by vedangmehta** [comment]
There are many topics of number theory including GCD, LCM, Sieve of Eratosthenes, Sieve of Atkin, prime factorization, Euler totient function, Miller-Rabin primality test that can be added to the number theory sections. It would be a great help to fellow programmer to avail Python implementation of all these algorithms at once.

**Comment by prakhar1989** [comment]
Okay. Makes sense. I'll create a section called "number theory" add this as well. The Miller-Rabin primality test already exists so that I'll love that accordingly. 

Let me know when you've added the `unittests` and I'll merge this. If you can, I would prefer that you [squash](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) your commits into one.

**Comment by vedangmehta** [comment]
You can currently add this algorithm to misc directory. Once we have enough Maths based algorithms, we can create a separate directory for this subtopic.

**Comment by prakhar1989** [comment]
Sure. Sounds good :+1:

File: Maths/sieve_of_eratosthenes.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    bool_array[1] = False
+    prime_array = []
+    upper_bound = ceil(sqrt(n))
+    for i in range(2,upper_bound):
+        if bool_array[i]:
+            for j in range(i*i,n+1,i):
+                bool_array[j] = False
+    for i in range(n+1):
+        if bool_array[i]:
+            prime_array.append(i)
```
Comment (prakhar1989) [inline Maths/sieve_of_eratosthenes.py]: You can use [list comprehensions](http://www.secnetix.de/olli/Python/list_comprehensions.hawk) to make this part of code much more succinct and pythonic.
```python
prime_array = []
for i in range(n+1):
     if bool_array[i]:
         prime_array.append(i)
return prime_array
```
Can be replaced with
```python
prime_array = [i for i in range(n+1) if bool_array[i]]
return prime_array 
```

If you're a beginner Python programmer than this *might* be slightly advanced but getting into the habit of using list comprehensions pays off great dividends later on :grinning:

**Comment by vedangmehta** [comment]
Thanks for your help. I just learnt about unit testing (also added a file to tests). Also changed the program as per your suggestion.

**Pull request closed without merging**